### PR TITLE
sql/mysql_db: Rework how we store and access table data for users, grants and replica_source_info.

### PR DESCRIPTION
**Pull request by reltuk**
In general, we keep the same structure of storing the data in structs, keyed in multimaps, and having converters for going to and from sql.Rows. We change the following important things:

1) We more explicitly model a multimap and an indexed set of entities with multiple keyers.

2) We add read and write locking around edits and reads of the data table.

3) We explicitly do not expose the raw indexed set or multimap from the MySQLDb instance itself. The unprincipled access of the various *Data instances in the old implementation was somewhat problematic.

### CODE REVIEW

File: sql/plan/grant.go [hunk 1]
```diff
@@ -664,33 +666,30 @@ func (n *GrantRole) CheckPrivileges(ctx *sql.Context, opChecker sql.PrivilegedOp
 	//TODO: only active roles may be assigned if the SUPER privilege is not held
 	mysqlDb := n.MySQLDb.(*mysql_db.MySQLDb)
 	client := ctx.Session.Client()
-	user := mysqlDb.GetUser(client.User, client.Address, false)
+
+	// TODO: Locking
```
Comment (fulghum) [inline sql/plan/grant.go:670]: Is there more locking needed here beyond what the reader encapsulates? I didn't understand concretely what this TODO was recommending.

**Review by fulghum — APPROVED** [review]
Nice refactoring! I only had a couple very minor thoughts. Looks like a nice improvement to the mysqldb.

File: sql/plan/grant.go (hunk 1 above)
Comment (reltuk) [reply sql/plan/grant.go:670]: This was a stale comment from before we had locking! Good catch :D.

File: sql/analyzer/privileges.go [hunk 2]
```diff
@@ -28,17 +28,25 @@ import (
 // TODO: add the remaining statements that interact with the grant tables
 func validatePrivileges(ctx *sql.Context, a *Analyzer, n sql.Node, scope *plan.Scope, sel RuleSelector) (sql.Node, transform.TreeIdentity, error) {
 	mysqlDb := a.Catalog.MySQLDb
+
+	// TODO: The management around closing this editor is sad. If you add
+	// an early exit before the Close after the GetUser below, make sure to
+	// close this editor.
+	ed := mysqlDb.Editor()
```
[the remark this answers is not in the document]
Comment (reltuk) [reply sql/analyzer/privileges.go:35]: I'm doing something kinda gross here where I'm using the `Editor()` to lock the access to `Enabled` as well. I will poke around to see how to change `Enabled` somehow to be more thread-safe.

**Pull request merged**